src/main/java/org/springframework/data/couchbase/repository/query/StringBasedN1qlQueryParser.java

```diff
@@ -109,36 +118,60 @@
 	private final CouchbaseConverter couchbaseConverter;
 	private final Collection<String> parameterNames = new HashSet<String>();
 	public final N1QLExpression parsedExpression;
 
-	public StringBasedN1qlQueryParser(String statement, QueryMethod queryMethod, String bucketName,
+	private GenericConversionService conversionService = new DefaultConversionService();
+	private CustomConversions conversions = new CouchbaseCustomConversions(Collections.emptyList());
+
+	public StringBasedN1qlQueryParser(String statement, CouchbaseQueryMethod queryMethod, String bucketName,
 			CouchbaseConverter couchbaseConverter, String typeField, String typeValue, ParameterAccessor accessor,
 			SpelExpressionParser parser, QueryMethodEvaluationContextProvider evaluationContextProvider) {
 		this.statement = statement;
 		this.queryMethod = queryMethod;
-		this.statementContext = createN1qlSpelValues(bucketName, typeField, typeValue, false);
-		this.countContext = createN1qlSpelValues(bucketName, typeField, typeValue, true);
 		this.couchbaseConverter = couchbaseConverter;
+		this.statementContext = createN1qlSpelValues(bucketName, null, null, typeField, typeValue, false, null);
+		this.countContext = createN1qlSpelValues(bucketName, null, null, typeField, typeValue, true, null);
 		this.parsedExpression = getExpression(accessor, getParameters(accessor), null, parser, evaluationContextProvider);
-		checkPlaceholders( this.parsedExpression.toString() );
+		checkPlaceholders(this.parsedExpression.toString());
+	}
+
+	public StringBasedN1qlQueryParser(String bucketName, CouchbaseConverter couchbaseConverter, Class domainClass,
+			Class resultClass, String typeField, String typeValue, String[] distinctFields) {
+		this.statement = null;
+		this.queryMethod = null;
+		this.couchbaseConverter = couchbaseConverter;
+		this.statementContext = createN1qlSpelValues(bucketName, domainClass, resultClass, typeField, typeValue, false,
+				distinctFields);
+		this.countContext = createN1qlSpelValues(bucketName, domainClass, resultClass, typeField, typeValue, true,
+				distinctFields);
+		this.parsedExpression = null;
 	}
 
-	public static N1qlSpelValues createN1qlSpelValues(String bucketName, String typeField, String typeValue,
-			boolean isCount) {
+	public N1qlSpelValues createN1qlSpelValues(String bucketName, Class domainClass, Class resultClass, String typeField,
+			String typeValue, boolean isCount, String[] distinctFields) {
 		String b = "`" + bucketName + "`";
+		String projectedFields = getProjectedFields(b, resultClass);
 		String entity = "META(" + b + ").id AS " + SELECT_ID + ", META(" + b + ").cas AS " + SELECT_CAS;
 		String count = "COUNT(*) AS " + CountFragment.COUNT_ALIAS;
 		String selectEntity;
-		if (isCount) {
+		if (distinctFields != null) {
+			String distinctFieldsStr = distinctFields.length == 0 ? projectedFields : getDistinctFields(distinctFields);
+			if (isCount) {
+				// only works with one field - MB43475
```

Comment:
can we raise an explicit exception to the user if its more than one field?

Reply:
This was my mistake.  The syntax for a sequence in n1ql is  DISTINCT { prop1, prop2 ... }
I was trying to use SQL syntax of DISTINCT prop1, prop2 ...
I've closed MB43475 and will fix my code.